src/source/src/dml_manager.rs

```diff
@@ -44,57 +52,190 @@ impl DmlManager {
         }
     }
 
+    /// Register a new DML reader for a table. If the reader for this version of the table already
+    /// exists, returns a reference to the existing reader.
     pub fn register_reader(
         &self,
         table_id: TableId,
+        table_version_id: TableVersionId,
         column_descs: &[ColumnDesc],
     ) -> Result<TableDmlHandleRef> {
         let mut table_readers = self.table_readers.write();
 
         // Clear invalid table readers.
-        table_readers.drain_filter(|_, weak_ref| weak_ref.strong_count() == 0);
-
-        match table_readers.entry(table_id) {
-            Entry::Occupied(o) => o.get().upgrade().ok_or_else(|| {
-                InternalError(format!(
-                    "fail to register reader for table with id {:?}",
-                    table_id.table_id
-                ))
-                .into()
-            }),
-            Entry::Vacant(v) => {
-                let reader = Arc::new(TableDmlHandle::new(column_descs.to_vec()));
-                v.insert(Arc::downgrade(&reader));
-                Ok(reader)
-            }
+        table_readers.drain_filter(|_, r| r.handle.strong_count() == 0);
+
+        macro_rules! new_handle {
+            ($entry:ident) => {{
+                let handle = Arc::new(TableDmlHandle::new(column_descs.to_vec()));
+                $entry.insert(TableReader {
+                    version_id: table_version_id,
+                    handle: Arc::downgrade(&handle),
+                });
+                handle
+            }};
         }
+
+        let handle = match table_readers.entry(table_id) {
+            // Create a new reader. This happens when the first `DmlExecutor` of this table is
+            // activated on this compute node.
+            Entry::Vacant(v) => new_handle!(v),
+
+            Entry::Occupied(mut o) => {
+                let TableReader { version_id, handle } = o.get();
+
+                match table_version_id.cmp(version_id) {
+                    // This should never happen as the schema change is guaranteed to happen after a
+                    // table is successfully created and all the readers are registered.
+                    Ordering::Less => unreachable!("table version `{table_version_id}` expired"),
+
+                    // Register with the correct version. This happens when the following
+                    // `DmlExecutor`s of this table is activated on this compute
+                    // node.
+                    Ordering::Equal => handle.upgrade().with_context(|| {
+                        format!("fail to register reader for table with key `{table_id:?}`")
+                    })?,
+
+                    // A new version of the table is activated, overwrite the old reader.
+                    Ordering::Greater => new_handle!(o),
```

Comment:
Will this lead to memory leak? There might be multiple old readers (old DML executors) holding the old `handle`, so the old `TableReader` cannot be dropped at this time because the ref count of `handle` not 0. There is however no way to reach the old `TableReader` after this overwriting, and it won't get dropped after the schema change completes. Maybe we can wrap the whole `TableReader` with `Weak` and directly return a `TableReader`. 🤔

Reply:
> and it won't get dropped after the schema change completes.

After schema change completes, the old DML executors will be all dropped and the `TableReader` itself should be dropped correctly. 🤔 Recall that it's the later `register_reader` that requires the manager to hold a (weak) ref to the reader. In this case, there should never be other registers with the old version, so dropping the weak reference makes sense.